clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -551,14 +556,19 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
             log.debug("Not returning fetched records for assigned partition {} since it is no longer fetchable",
                     partitionRecords.partition);
         } else {
-            long position = subscriptions.position(partitionRecords.partition);
-            if (partitionRecords.nextFetchOffset == position) {
+            SubscriptionState.FetchPosition position = subscriptions.position(partitionRecords.partition);
+            if (partitionRecords.nextFetchOffset == position.offset) {
                 List<ConsumerRecord<K, V>> partRecords = partitionRecords.fetchRecords(maxRecords);
 
-                long nextOffset = partitionRecords.nextFetchOffset;
-                log.trace("Returning fetched records at offset {} for assigned partition {} and update " +
-                        "position to {}", position, partitionRecords.partition, nextOffset);
-                subscriptions.position(partitionRecords.partition, nextOffset);
+                if (partitionRecords.nextFetchOffset > position.offset) {
```

Comment:
I am sure I am missing something but how can this be true when due to line 560 `if (partitionRecords.nextFetchOffset == position.offset) {`?

Reply:
On a side note, unless the local variable and the field are marked as `final`, it is basically impossible for javac to catch this statically because Java's concurrency model. E.g. any thread can modify both `partitionRecords.nextFetchOffset` and `position.offset`

Reply:
Good catch. Let me look into this... might be something I missed during the merge from trunk.

Reply:
I think the code is correct as-is. The call to `partitionRecords.fetchRecords(maxRecords)` updates the nextFetchOffset to the offset of the last record returned plus one.